virtcontainers/qemu.go

```diff
@@ -929,6 +933,58 @@ func (q *qemu) hotplugAddMemory(memDev *memoryDevice) error {
 	return q.sandbox.storage.storeHypervisorState(q.sandbox.id, q.state)
 }
 
+func (q *qemu) hotplugVSock(op operation) (uint32, error) {
+	if op == removeDevice {
+		return 0, errors.New("Remove vsock is not supported, it's removed once sandbox have finished")
+	}
+
+	return q.hotplugAddVSock()
+}
+
+func (q *qemu) hotplugAddVSock() (uint32, error) {
+	var err error
+	var cid uint32
+
+	// setup qmp channel if necessary
+	if q.qmpMonitorCh.qmp == nil {
+		qmp, err := q.qmpSetup()
+		if err != nil {
+			return 0, err
+		}
+
+		q.qmpMonitorCh.qmp = qmp
+
+		defer func() {
+			qmp.Shutdown()
+			q.qmpMonitorCh.qmp = nil
+		}()
+	}
+
+	// ExecutePCIVSockAdd fails if the context ID was used for other process just before
+	// adding the vsock, we have to find other context ID and try again.
+	for i := 0; i < hotplugVSockMaxTries; i++ {
```

Comment:
I'm confused about the need for a retry here. I thought that using `VHOST_VSOCK_SET_GUEST_CID` ioctl on `/dev/vhost-vsock` would reserve the CID so that nobody else can use it. I have found this [comment](https://bugzilla.redhat.com/show_bug.cgi?id=1291851#c6) from @stefanha (vsock maintainer), and I think he doesn't mention that the CID could be stolen after it's been properly set.
Let me know but I feel like this could be simplified by calling `findContextID()` only once.

Reply:
I don't think so 

see https://elixir.bootlin.com/linux/latest/source/drivers/vhost/vsock.c#L608 and https://elixir.bootlin.com/linux/latest/source/drivers/vhost/vsock.c#L670

```C
static int vhost_vsock_set_cid(struct vhost_vsock *vsock, u64 guest_cid)
{
	struct vhost_vsock *other;

	/* Refuse reserved CIDs */
	if (guest_cid <= VMADDR_CID_HOST ||
	    guest_cid == U32_MAX)
		return -EINVAL;

	/* 64-bit CIDs are not yet supported */
	if (guest_cid > U32_MAX)
		return -EINVAL;

	/* Refuse if CID is already in use */
	spin_lock_bh(&vhost_vsock_lock);
	other = __vhost_vsock_get(guest_cid);
	if (other && other != vsock) {
		spin_unlock_bh(&vhost_vsock_lock);
		return -EADDRINUSE;
	}
	vsock->guest_cid = guest_cid;
	spin_unlock_bh(&vhost_vsock_lock);

	return 0;
}
```

Reply:
We took the discussion offline and agreed this can be handled without any race condition if we hold onto the file descriptor and if we pass it to the qemu command line.